Update and re-enable the diagnostics for unstable runtime names

jrose-apple opened the pull request:
(builds on #10078)

This time, the warnings only fire when the class in question directly conforms to NSCoding. This avoids warning on cases where the user has subclassed something like, oh, UIViewController, and has no intention of writing it to a persistent file.

This also removes the warning for generic classes that conform to NSCoding, for simplicity's sake. That means `@NSKeyedArchiverEncodeNonGenericSubclassesOnly` is also being removed.

Actually archiving a class with an unstable mangled name is still considered problematic, but the compiler shouldn't emit diagnostics unless it can be sure they are relevant.

rdar://problem/32314195

jrose-apple committed:
Allow '@objc(RuntimeName)' on classes with generic ancestry.

This is accomplished by recognizing this specific situation and
replacing the 'objc' attribute with a hidden '_objcRuntimeName'
attribute. This /only/ applies to classes that are themselves
non-generic (including any enclosing generic context) but that have
generic ancestry, and thus cannot be exposed directly to Objective-C.

This commit also eliminates '@NSKeyedArchiverClassName'. It was
decided that the distinction between '@NSKeyedArchiverClassName' and
'@objc' was too subtle to be worth explaining to developers, and that
any case where you'd use '@NSKeyedArchiverClassName' was already a
place where the ObjC name wasn't visible at compile time.

This commit does not update diagnostics to reflect this change; we're
going to change them anyway.

rdar://problem/32414557

jrose-apple commented:
cc @amartini51

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please test macOS

jrose-apple commented:
@swift-ci Please smoke test Linux

jrose-apple commented:
@swift-ci Please test macOS

DougGregor commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -1481,10 +1481,36 @@ bool Parser::parseDeclAttribute(DeclAttributes &Attributes, SourceLoc AtLoc) {
-    diagnose(Tok, diag::attr_nskeyedarchiverclassname_removed)
-      .fixItReplace(Tok.getLoc(), "objc");
+    auto activeDiag = diagnose(Tok,diag::attr_nskeyedarchiverclassname_removed);
+    activeDiag.fixItReplace(Tok.getLoc(), "objc");
```
Ah, I see you corrected the Fix-It here that I complained about in the previous PR. Thanks!

DougGregor reviewed (approved):
LGTM. As with the previous PR, this is a much simpler scheme.

itaiferber reviewed (approved):
Looks reasonable to me! Thanks for putting this together!

jrose-apple committed:
Update and re-enable the diagnostics for unstable runtime names.

This time, the warnings only fire when the class in question directly
conforms to NSCoding. This avoids warning on cases where the user has
subclassed something like, oh, UIViewController, and has no intention
of writing it to a persistent file.

This also removes the warning for generic classes that conform to
NSCoding, for simplicity's sake. That means
'@NSKeyedArchiverEncodeNonGenericSubclassesOnly' is also being
removed.

Actually archiving a class with an unstable mangled name is still
considered problematic, but the compiler shouldn't emit diagnostics
unless it can be sure they are relevant.

rdar://problem/32314195

jrose-apple commented:
@swift-ci Please test
